transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java

```diff
@@ -125,62 +125,83 @@ protected void run() {
-                try {
-                    if (curDeadlineNanos != prevDeadlineNanos) {
-                        prevDeadlineNanos = curDeadlineNanos;
-                        submissionQueue.addTimeout(deadlineToDelayNanos(curDeadlineNanos));
-                        submissionQueue.submit();
-                    }
+                // Prepare to block wait
+                long curDeadlineNanos = nextScheduledTaskDeadlineNanos();
+                if (curDeadlineNanos == -1L) {
+                    curDeadlineNanos = NONE; // nothing on the calendar
+                }
+                nextWakeupNanos.set(curDeadlineNanos);
 
-                    // Check there were any completion events to process
-                    if (completionQueue.process(this) == -1) {
-                        // Block if there is nothing to process after this try again to call process(....)
-                        logger.trace("ioUringWaitCqe {}", this.toString());
-                        completionQueue.ioUringWaitCqe();
+                // Only submit a timeout if there are no tasks to process and do a blocking operation
+                // on the completionQueue.
+                try {
+                    if (!hasTasks()) {
+                        if (curDeadlineNanos != prevDeadlineNanos) {
+                            prevDeadlineNanos = curDeadlineNanos;
+                            submissionQueue.addTimeout(deadlineToDelayNanos(curDeadlineNanos));
+                        }
+
+                        // Check there were any completion events to process
+                        if (!completionQueue.hasCompletions()) {
+                            // Block if there is nothing to process after this try again to call process(....)
+                            logger.trace("submitAndWait {}", this);
+                            submissionQueue.submitAndWait();
+                        }
                     }
-                } catch (Throwable t) {
-                    //Todo handle exception
                 } finally {
                     if (nextWakeupNanos.get() == AWAKE || nextWakeupNanos.getAndSet(AWAKE) == AWAKE) {
                         pendingWakeup = true;
                     }
                 }
-            }
-
-            completionQueue.process(this);
-
-            // Always call runAllTasks() as it will also fetch the scheduled tasks that are ready.
-            runAllTasks();
 
-            submissionQueue.submit();
-            try {
-                if (isShuttingDown()) {
-                    closeAll();
-                    submissionQueue.submit();
-
-                    if (confirmShutdown()) {
-                        break;
+                // avoid blocking for as long as possible - loop until available work exhausted
+                boolean maybeMoreWork = true;
+                do {
+                    try {
+                        maybeMoreWork = completionQueue.process(this) != 0 | runAllTasks();
```

Comment:
I'm not sure if a Or bitwise operator would make sense here, what about `||`?

Reply:
the difference here is (IMHO) that with `|` it will run both commands while with `||` it will run only the first. If `|` is really what we want we should at least clarify this in a comment as at the first look it seems suspicious.

Reply:
okay that makes sense :)

Reply:
Yep agreed a comment would be good!